### PR TITLE
The --info option produces a list of used functions and distributions

### PR DESCRIPTION
**Pull request by mandel**
## Release notes

As part of the issue #808, it was discuss to add the list of  "all named functions used". I understood that as the list for functions called from the standard library and the list of used distribution.

This PR adds to the JSON object produced by the `--info` option the two following fields:
- `stanlib_calls`: list of functions of the standard library called in the program
- `distributions`: list of distributions to the right of a `~` statement.

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by mandel** [comment]
I do not understand why the changes in this PR cause the following error:
```
--- Translating Stan model to C++ code ---

bin/stanc  --o=examples/bernoulli/bernoulli.hpp examples/bernoulli/bernoulli.stan


--- Compiling, linking C++ code ---

clang++-6.0 -march=core2 -std=c++1y -D_REENTRANT -Wno-sign-compare -Wno-ignored-attributes      -I stan/lib/stan_math/lib/tbb_2019_U8/include   -O3 -I src -I stan/src -I lib/rapidjson_1.1.0/ -I lib/CLI11-1.9.1/ -I stan/lib/stan_math/ -I stan/lib/stan_math/lib/eigen_3.3.9 -I stan/lib/stan_math/lib/boost_1.75.0 -I stan/lib/stan_math/lib/sundials_5.6.1/include    -DBOOST_DISABLE_ASSERTS         -c -include-pch stan/src/stan/model/model_header.hpp.gch -x c++ -o examples/bernoulli/bernoulli.o examples/bernoulli/bernoulli.hpp

fatal error: file '/usr/include/linux/futex.h' has been modified since the precompiled header 'stan/src/stan/model/model_header.hpp.gch' was built
```

It looks unrelated to the `--info` option.

**Comment by rok-cesnovar** [comment]
This is unrelated, maybe a one-off error on our test machine or a configuration issue on the machine. In both cases, dont worry about that, we will look into that.

Thanks for the PR! Looks nice.

Two comments/questions:
- I think we dont need to separate distributions? Or do you think there is a use case? Might be easier to just list `dirichlet_lpdf` in stanlib_calls? You  can call this for all tilde statements: 
https://github.com/stan-dev/stanc3/blob/master/src/middle/Utils.ml#L48

- We probably need to add a test for higher-order functions. Example models to use: https://github.com/stan-dev/stanc3/blob/master/test/integration/good/code-gen/reduce_sum_m1.stan or https://github.com/stan-dev/stanc3/blob/master/test/integration/good/ode/ode_rk45_test_model.stan

**Comment by mandel** [comment]
> I think we dont need to separate distributions? Or do you think there is a use case?

I do not have a strong opinion but it is easier to add a post-processing phase that merge them than the converse. Moreover, in our Stan to (Num)Pyro compiler with keep the two forms separated and might have a benefit to have access to this information.

> Might be easier to just list dirichlet_lpdf in stanlib_calls? You can call this for all tilde statements: https://github.com/stan-dev/stanc3/blob/master/src/middle/Utils.ml#L48

It seems that this function does not add the `_lpdf` suffix when called on `"dirichlet"`.

**Comment by mandel** [comment]
> We probably need to add a test for higher-order functions. Example models to use: https://github.com/stan-dev/stanc3/blob/master/test/integration/good/code-gen/reduce_sum_m1.stan or https://github.com/stan-dev/stanc3/blob/master/test/integration/good/ode/ode_rk45_test_model.stan

That is an interesting question. In these examples, the functions in argument of `reduce_sum` and `ode_rk45` are local functions so not reported anyway. But if the function in argument is a function from the Stan standard library it would currently not be present in the list `stanlib_calls`. Would you prefer a different behavior?

**Comment by seantalts** [comment]
> I do not have a strong opinion but it is easier to add a post-processing phase that merge them than the converse. 

+1, I think it's neat to present additional information that could be helpful. I'm not sure if this is better or worse but we could also structure it something like
`functions_called : [{"dirichlet": {"distribution": true}, "log": {"distribution": false}]`

**Comment by rok-cesnovar** [comment]
> I do not have a strong opinion but it is easier to add a post-processing phase that merge them than the converse. 

Very good point. Ignore my previous comment. Splitting these up is good.

>  the functions in argument of reduce_sum and ode_rk45 are local functions so not reported anyway.

Ah yes, I was overthinking this. Obviously these call user-defined functions so the current behaviour is fine. We were discussing functors for Stan lib functions in https://github.com/stan-dev/design-docs/pull/35 but we can address that if we actually do start supporting it.

File: test/integration/cli-args/info.stan [hunk 1]
```diff
@@ -29,12 +29,13 @@ parameters {
...
   {
-    real ignored_tp = r;
+    real ignored_tp = log(r[0][0]);
   }
 }
 
 model {
-    real ignored_model = r;
+    real ignored_model = square(r[0][0]);
+    l ~ dirichlet(c);
```
Comment (rok-cesnovar) [inline test/integration/cli-args/info.stan:38]:
```suggestion
    l ~ dirichlet(c);
    target += normal_lpdf(y| 0, 1);
    target += normal_lcdf(y| 0, 1);
    target += normal_lccdf(y| 0, 1);
```

Also requires adding `real y;` in the parameters.

The functions used in `target +=` are not listed.

**Review by rok-cesnovar — COMMENTED** [review]
Thanks again for working on this!

Sorry I suggested the wrong function in my comment. This is the example (we only care of the StanLib ones):
```ocaml
let possible_names =
    List.map ~f:(( ^ ) distribution.name) Utils.distribution_suffices
    |> String.Set.of_list
in
if List.exists ~f:(fun (n, _) -> Set.mem possible_names n) ud_dists (* ud_dists is the 'dirichlet' in our case *)
then Fun_kind.UserDefined
else StanLib
```

I am nitpicking here because I think 
```
    l ~ dirichlet(c);
    target += normal_lpdf(y| 0, 1);
```
should both return functions either with or without the suffix. I prefer with suffix but would be fine without it as well as long as its consistent.

Also please add a test for Stan lib functions in user-defined functions:

```stan
functions {
  real foo(real a) {
    return sin(a);
  }
  real goo_lpdf(real a) {
    return a;
  }
}
```

If I then add 
```stan
    y ~ goo();
```
goo should not be listed.

**Comment by mandel** [comment]
Thanks @rok-cesnovar. I have implemented most of your suggestions. But for now we still distinguish distributions with and without post-fix. 

If you do not like the current solution, I also prefer the solution where the post-fix is added since it is always easy to remove the post-fix later. 

But in this case, I think that it would be good to expose the function `dist_name_suffix` (https://github.com/stan-dev/stanc3/blob/master/src/frontend/Ast_to_Mir.ml#L501-L511) such that this code is not duplicated. I don't know if it should be added to the interface of `Ast_to_Mir` or moved to `Middle.Utils`?

**Comment by rok-cesnovar** [comment]
Thanks!

I think we want to have the _lpdf/_lpmf suffixes. My reasoning is that these two models are equivalent, produce the exact same C++ code and should also produce the same output with --info.
```stan
parameters {
    real y;
}
model {
    y ~ std_normal();
}

parameters {
    real y;
}
model {
    target += std_normal_lpdf(y);
}
```

> But in this case, I think that it would be good to expose the function dist_name_suffix (https://github.com/stan-dev/stanc3/blob/master/src/frontend/Ast_to_Mir.ml#L501-L511) such that this code is not duplicated. I don't know if it should be added to the interface of Ast_to_Mir or moved to Middle.Utils?

I would say Middle.Utils but not sure. Do you have a preference @seantalts ? Utils has all the other distribution suffix utilities, idk.

File: test/integration/cli-args/info.stan [hunk 2]
```diff
@@ -24,17 +33,23 @@ parameters {
...
   }
 }
 
 model {
-    real ignored_model = r;
+    real ignored_model = square(foo(square(r[0][0])));
+    l ~ dirichlet(c);
+    target += normal_lpdf(y| 0, 1);
+    target += normal_lcdf(y| 0, 1);
+    target += normal_lccdf(y| 0, 1);
```
Comment (rok-cesnovar) [inline test/integration/cli-args/info.stan:51]:
```suggestion
    target += normal_lccdf(y| 0, 1);
    target += std_normal_lupdf(y)
```

This one should be listed as `std_normal_lpdf` (target += lupdf is the same as using the sampling statement).

For this rename you can just call Utils.stdlib_distribution_name on all distributions.

**Review by rok-cesnovar — COMMENTED** [review]
My above comment got me thinking about another test case. Sorry I missed this earlier.

**Comment by mandel** [comment]
In fact, it is not possible to move `dist_name_suffix` in `Middle.Utils` because it introduces a cyclic dependency with `Middle.Stan_math_signatures`. So should it be in `Middle.Stan_math_signatures`?

**Comment by rok-cesnovar** [comment]
> So should it be in Middle.Stan_math_signatures?

I think so yes.

I would also rename `stanlib_calls` to `functions`. I think then its good to go.

**Review by rok-cesnovar — APPROVED** [review]
Thank you Louis!

**Comment by mandel** [comment]
Your welcome!

**Pull request merged**